### PR TITLE
bluestore: use string_view and bump rocksdb to v6.15.5

### PR DESCRIPTION
**Pull request by tchaikov**
to pick up a bunch of cleanups which silence warnings like range-based
loop using `auto` instead of `auto &`.

Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
this change fails the "make check" test.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by jdurgin** [comment]
@markhpc should we update our version of rocksdb's cache code that we've copied into bluestore as well?

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -3176,8 +3176,8 @@ void BlueFS::_maybe_compact_log(std::unique_lock<ceph::mutex>& l)
 }
 
 int BlueFS::open_for_write(
-  const string& dirname,
-  const string& filename,
+  string_view dirname,
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc:3179]: suggest to use (or do not use) std:: prefix uniformly everywhere through the patch

File: src/os/bluestore/BlueRocksEnv.cc [hunk 2]
```diff
@@ -440,7 +440,13 @@ rocksdb::Status BlueRocksEnv::GetChildren(
   std::vector<std::string>* result)
 {
   result->clear();
-  int r = fs->readdir(dir, result);
+  // dir may contain a trailing /
```
Comment (ifed01) [inline src/os/bluestore/BlueRocksEnv.cc:443]: wouldn't it be better to put this inside readdir?

Comment (tchaikov) [reply src/os/bluestore/BlueRocksEnv.cc]: indeed! pushed it down to `BlueFS::readdir()`.

**Review by tchaikov — COMMENTED** [review]
@ifed01 updated and repushed

File: src/os/bluestore/BlueFS.cc (hunk 1 above)
Comment (tchaikov) [reply src/os/bluestore/BlueFS.cc]: sure. now use `std::string_view` uniformly.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -16264,7 +16264,7 @@ void RocksDBBlueFSVolumeSelector::get_paths(const std::string& base, paths& res)
   res.emplace_back(base + ".slow", l_totals[LEVEL_SLOW - LEVEL_FIRST]);
 }
 
-void* RocksDBBlueFSVolumeSelector::get_hint_by_dir(const string& dirname) const {
+void* RocksDBBlueFSVolumeSelector::get_hint_by_dir(string_view dirname) const {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:16267]: still missing std::

Comment (tchaikov) [reply src/os/bluestore/BlueStore.cc:16267]: @ifed01 fixed and repushed.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-03-20_13:57:49-rados-wip-kefu2-testing-2021-03-20-1320-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-03-22_02:52:16-rados-wip-kefu2-testing-2021-03-21-2353-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/49916
- https://tracker.ceph.com/issues/49909

**Pull request merged**